File: task_scheduler.py
import tkinter as tk
from tkinter import ttk
from PIL import Image , ImageTk #Pour l'icon
import csv
from tkcalendar import DateEntry 
import subprocess



#Enregistrer dans le table cron
def enregistrer_nouvelle_tache(nouvelle_tache):

    tache_cron = ""
    type_planification = ["Exécution unique","Tous les jours","Chaque semaine","Chaque mois"]
    ##Ecrire les tâches sur le table cron
    #Lire le crontab actuel
    resultat = subprocess.run(['crontab','-l'],capture_output=True,text=True)
    crontab_actuel=resultat.stdout
    #Ajouter une nouvelle tâche 
        #Forme de tâche cron : 
        # * * * * * commande
        # │ │ │ │ │
        # │ │ │ │ └── Jour de la semaine (0-7) [0 et 7 = dimanche]
        # │ │ │ └──── Mois (1-12)
        # │ │ └────── Jour du mois (1-31)
        # │ └──────── Heure (0-23)
        # └────────── Minute (0-59)
    
    h,min=nouvelle_tache[4].split(":")
    anne,mois,jour=str(nouvelle_tache[3]).split("-")
    #Forme de chaque tache dans le table cron 
    #   * * * * * commande      #Nom de la tâche:Type de planification
    if nouvelle_tache[2]==type_planification[0]:
        tache_cron =min + " " + h + " " + jour + " " + mois + " " +  "*" + " " + nouvelle_tache[1] + "\t#" + nouvelle_tache[0] + ":" + nouvelle_tache[2] + "\n"  
    elif nouvelle_tache[2] == type_planification[1]:
        tache_cron = min + " " + h + " " +  "*" + " " + "*" + " "+ "*" + " " + nouvelle_tache[1] + "\t#" + nouvelle_tache[0] + ":" + nouvelle_tache[2] + "\n"
    elif nouvelle_tache[2] == type_planification[2]:
        tache_cron = min + " " + h + " " + "*" + " " + "*" + " " + nouvelle_tache[6] + " " + nouvelle_tache[1] + "\t#" + nouvelle_tache[0] + ":" + nouvelle_tache[2] + "\n"  
    elif nouvelle_tache[2] == type_planification[3]:
        tache_cron =min + " " + h + " " +  nouvelle_tache[5] + " " + "*" + " " + "*" + " " + nouvelle_tache[1] + "\t#" + nouvelle_tache[0] + ":" + nouvelle_tache[2] + "\n"  
    nouveau_crontab = crontab_actuel + tache_cron

    #Écrire le nouveau crontab 
    subprocess.run(['crontab','-'],input=nouveau_crontab,text=True)

def ajouter_une_tache(*args):
    retoure=None
    tache=["","","","00/00/00","00:00","*","*"]
    jours=["Dimanche","Lundi","Mardi","Mercredi","Jeudi","Vendredi","Samedi"]
    type_planification = ["Exécution unique","Tous les jours","Chaque semaine","Chaque mois"]
    fenetre = tk.Tk()
    if args :
        fenetre.title(args[0])
    else: 
        fenetre.title("Nouvelle tâche")
    fenetre.geometry("500x275")
    fenetre.columnconfigure(1, weight=1)
    #Nom de la tâche
    tk.Label(fenetre, text="Nom de la tâche").grid(row=0, column=0, sticky="w", padx=10, pady=5)
    nom_entry = tk.Entry(fenetre)
    nom_entry.grid(row=0, column=1, sticky="we", padx=10, pady=5)
    #Commande  à exécuter
    tk.Label(fenetre, text="Commande").grid(row=1, column=0, sticky="w", padx=10, pady=5)
    commande_entry = tk.Entry(fenetre)
    commande_entry.grid(row=1, column=1, sticky="we", padx=10, pady=5)
    #Type de planification
    tk.Label(fenetre, text="Type de planification").grid(row=2, column=0, sticky="w", padx=10, pady=5)
    type=ttk.Combobox(fenetre,values=type_planification)
    type.grid(row=2, column=1, sticky="we", padx=10, pady=5)
    #Date
    tk.Label(fenetre, text="Date").grid(row=3, column=0, sticky="w", padx=10, pady=5)
    date_entry = DateEntry(fenetre, date_pattern="dd/mm/yyyy")
    date_entry.grid(row=3, column=1, sticky="we", padx=10, pady=5)
    #Heure
    tk.Label(fenetre, text="Heure").grid(row=4, column=0, sticky="w", padx=10, pady=5)
    heure = tk.Spinbox(fenetre, from_=0, to=23, width=5)
    heure.grid(row=4, column=0, sticky="e", padx=10)
    minute = tk.Spinbox(fenetre, from_=0, to=59, width=5)
    minute.grid(row=4, column=1, sticky="w", padx=10)
    #Jour du mois
    tk.Label(fenetre, text="Jour du mois").grid(row=5, column=0, sticky="w", padx=10, pady=5)
    jour_mois = tk.Spinbox(fenetre, from_=0, to=31, width=5)
    jour_mois.grid(row=5, column=0, sticky="e", padx=10)
    #Jour de la semaine
    tk.Label(fenetre, text="Jour de la semaine").grid(row=6, column=0, sticky="w", padx=10, pady=5)
    jour_semaine=ttk.Combobox(fenetre,values=jours)
    jour_semaine.grid(row=6, column=1, sticky="we", padx=10, pady=5)
    
    def annuler():
        nonlocal retoure
        retoure=-1
        fenetre.destroy()

    #Recuperer toutes les informations sur la tâche 
    def enregistrer():
        nonlocal retoure
        tache[0] = nom_entry.get()
        tache[1] = commande_entry.get()
        tache[2] = type.get()
        tache[3] = date_entry.get_date()
        tache[4] = heure.get() + ":" + minute.get()
        tache[5] = jour_mois.get()
        for i, jour in enumerate(jours):
            if jour == jour_semaine.get():
                tache[6] = str(i)
                break
        enregistrer_nouvelle_tache(tache)
        retoure = 1
        fenetre.destroy()


    #Boutons Annuler , Enregistrer
    btn_annuler=tk.Button(fenetre, text="Annuler",bg="#222121",width=25,fg="white",command=annuler)
    btn_annuler.grid(row=8,column=0,sticky='w')
    btn_enregistrer=tk.Button(fenetre, text="Enregistrer",bg="#222121",width=25,fg="white",command=enregistrer)
    btn_enregistrer.grid(row=8,column=1,sticky='e')

    fenetre.mainloop()
    return(retoure)


def recuperer_tache_cron_dans_fichier():
    jours = ["Dimanche","Lundi","Mardi","Mercredi","Jeudi","Vendredi","Samedi"]
    type_planification = ["Exécution unique","Tous les jours","Chaque semaine","Chaque mois"]
    
    # Lire le crontab actuel
    resultat = subprocess.run(['crontab', '-l'], capture_output=True, text=True)
    lignes = resultat.stdout.splitlines()
    
    taches = []
    
    for ligne in lignes:
        # Ignorer les lignes vides ou commentaires purs
        if not ligne.strip() or ligne.strip().startswith("#"):
            continue
        
        # Séparer la partie cron du commentaire  (#Nom:Type)
        if "#" in ligne:
            partie_cron, commentaire = ligne.split("#", 1)
        else:
            continue  # Sans commentaire, on ne peut pas récupérer le nom/type
        
        partie_cron = partie_cron.strip().split()
        if len(partie_cron) < 6:
            continue
        
        # Extraire les champs cron : minute heure jour_mois mois jour_semaine commande
        minute       = partie_cron[0]
        heure        = partie_cron[1]
        jour_mois    = partie_cron[2]
        mois         = partie_cron[3]
        jour_semaine = partie_cron[4]
        commande     = " ".join(partie_cron[5:])
        
        # Extraire nom et type depuis le commentaire
        if ":" in commentaire:
            nom, type_tache = commentaire.split(":", 1)
            nom = nom.strip()
            type_tache = type_tache.strip()
        else:
            nom = commentaire.strip()
            type_tache = ""
        
        # Reconstruire date et heure
        heure_exec = heure.zfill(2) + ":" + minute.zfill(2)
        
        if type_tache == type_planification[0]:  # Exécution unique
            date_exec = f"20{mois.zfill(2)}-{mois.zfill(2)}-{jour_mois.zfill(2)}"
            # Reconstruire : année inconnue, on met les champs disponibles
            date_exec = f"??/{mois.zfill(2)}/{jour_mois.zfill(2)}"
        else:
            date_exec = "00/00/00"
        
        taches.append([nom, commande, type_tache, date_exec, heure_exec, jour_mois, jour_semaine])
    
    # Écrire dans liste_taches.csv
    with open("liste_taches.csv", "w", newline="", encoding="utf-8") as fichier:
        writer = csv.writer(fichier)
        if taches:
            writer.writerows(taches)
        else:
            # Ligne par défaut si aucune tâche
            writer.writerow(["", "", "", "00/00/00", "00:00", "*", "*"])



def liste_de_taches(centre_frame):
    recuperer_tache_cron_dans_fichier()
    jours=["Dimanche","Lundi","Mardi","Mercredi","Jeudi","Vendredi","Samedi"]
    type_planification=["Exécution unique","Tous les jours","Chaque semaine","Chaque mois"]

    # Canvas
    canvas = tk.Canvas(centre_frame,bg="#4D4C4C")
    canvas.pack(side=tk.LEFT, fill="both", expand=True)
    # Barre de défilement
    scrollbar = tk.Scrollbar(centre_frame, orient="vertical", command=canvas.yview)
    scrollbar.pack(side="right", fill="y")
    canvas.configure(yscrollcommand=scrollbar.set)
    # Frame contenant les tâches
    conteneur = tk.Frame(canvas,bg="#4D4C4C")
    canvas.create_window((0, 0), window=conteneur, anchor="nw")
    conteneur.bind(
        "<Configure>",
        lambda e: canvas.configure(scrollregion=canvas.bbox("all"))
    )

    ##Lire le fichier liste_taches.csv qui contient toutes les tâches dans crontab -l avec leurs informations
    #Forme de liste_taches.csv : Nom de tâche,Commande à exécuter,Type de planification,Date d'exécution ,Heure d'exécution,Jour du mois , Jour de la semaine
    with open("liste_taches.csv") as fichier:
        donnes = csv.reader(fichier)
        for ligne in donnes:
            liste_frame=tk.Frame(conteneur,bg="#4D4C4C",height=80,highlightthickness=3,highlightbackground="#616164")
            liste_frame.pack(fill=tk.X,padx=50,pady=5,ipadx=310,ipady=5)
            liste_frame.pack_propagate(False)
            if ligne[2]==type_planification[0]:
                date=ligne[3]+" à "+ligne[4]
            elif ligne[2]==type_planification[1]:
                date=ligne[2]+" à "+ligne[4]
            elif ligne[2]==type_planification[2]:
                date=ligne[2]+": Tous le "+jours[int(ligne[6])] + " à " + ligne[4]
            elif ligne[2]==type_planification[3]:
                date=ligne[2]+" : Tous le "+ligne[5] + "ème jour du mois à " + ligne[4]
            commande="("+ligne[1]+")"
            #Afficher le nom , commande et date d'exécution de chque tâche 
            tache_label=tk.Label(liste_frame,text=ligne[0],font=("TkDefaultFont",12,"bold"),bg="#4D4C4C")
            tache_label.pack(fill=tk.X)
            commande_label=tk.Label(liste_frame,text=commande,fg="#AFA2A2",bg="#4D4C4C")
            commande_label.pack(fill=tk.X)
            date_label=tk.Label(liste_frame,text=date,fg="#FFFFFF",bg="#4D4C4C")
            date_label.pack(fill=tk.X)

            #Boutton pour supprimer et modifier 
            icon=ImageTk.PhotoImage(Image.open("delete.png").resize((25,25)))
            supprimer=tk.Button(liste_frame,image=icon ,width=25,height=3)
            supprimer.pack(side=tk.RIGHT,anchor='center')

#Cette fonction permet de supprimer une tâche qui contient la chaîne dans la variable "tache_a_supprimer" dans le table cron
def supprimer_une_tache_dans_table_cron(tache_a_supprimer):
    #Lire le crontab actuel
    resultat = subprocess.run(['crontab','-l'],capture_output=True,text=True)
    crontab_actuel=resultat.stdout

    #Supprimer une tâche 
    liste_tache = crontab_actuel.splitlines()
    for tache in liste_tache:
        if str(tache_a_supprimer) in tache :
            liste_tache.remove(tache)
    
    #Réecrire le nouveau crontab
    crontab_actuel="\n".join(liste_tache)+ "\n"
    subprocess.run(['crontab','-'],input=crontab_actuel,text=True)




def modifier_une_tache():
    fenetre = tk.Tk()
    fenetre.title("Nom de la tâhe")
    fenetre.geometry("350x120")
    nom_label=tk.Label(fenetre,text="Nom de la tâche / Commande ")
    nom_label.grid(row=1,column=0)
    nom_entry=tk.Entry(fenetre)
    nom_entry.grid(row=2, column=0, columnspan=2, sticky="we", padx=5,pady=10)
    tache=nom_entry.get()
    def annuler():
        fenetre.destroy()
    def enregistrer():
        fenetre.destroy()
        retoure=ajouter_une_tache("Modification d'une tâche")
        if retoure==1:
            supprimer_une_tache_dans_table_cron(tache)

    btn_annuler=tk.Button(fenetre, text="Annuler",bg="#222121",width=15,fg="white",command=annuler)
    btn_annuler.grid(row=3,column=0,sticky='w',pady=10)
    btn_enregistrer=tk.Button(fenetre, text="Modifier",bg="#222121",width=15,fg="white",command=enregistrer)
    btn_enregistrer.grid(row=3,column=1,sticky='e',pady=10)
    fenetre.mainloop()
    

#Fonction qui permet de supprimer une tâche 
def supprimer_une_tache():
    fenetre = tk.Tk()
    fenetre.title("Nom de la tâhe")
    fenetre.geometry("350x120")
    nom_label=tk.Label(fenetre,text="Nom de la tâche / Commande ")
    nom_label.grid(row=1,column=0)
    nom_entry=tk.Entry(fenetre)
    nom_entry.grid(row=2, column=0, columnspan=2, sticky="we", padx=5,pady=10)
    
    def annuler():
        fenetre.destroy()
    def enregistrer():
        supprimer_une_tache_dans_table_cron(nom_entry.get())
        fenetre.destroy()

    btn_annuler=tk.Button(fenetre, text="Annuler",bg="#222121",width=15,fg="white",command=annuler)
    btn_annuler.grid(row=3,column=0,sticky='w',pady=10)
    btn_enregistrer=tk.Button(fenetre, text="Supprimer",bg="#222121",width=15,fg="white",command=enregistrer)
    btn_enregistrer.grid(row=3,column=1,sticky='e',pady=10)
    fenetre.mainloop()
    



def menu_principale():
    fenetre = tk.Tk()
    fenetre.title("PLANIFICATEUR DE TÂCHE")
    fenetre.geometry("1000x700")

    #Créer le frame de menu
    menu_frame = tk.Frame(fenetre,bg="#222121",width=250,highlightthickness=3,highlightbackground="#616164")
    menu_frame.pack(side=tk.LEFT,fill=tk.Y)
    menu_frame.pack_propagate(False) #Fixe le largeur du frame

    #Creer "centre_frame"
    centre_frame = tk.Frame(fenetre,bg="#4D4C4C",highlightthickness=3,highlightbackground="#616164")
    centre_frame.pack(fill=tk.BOTH,expand=True)
    centre_frame.pack_propagate(False) #Fixe le largeur du frame

    ##Boutton : Nouvelle tâche
    #Creer un boutton "Nouvelle tâche" qui permet de créer une tâche
    icon1=ImageTk.PhotoImage(Image.open("add.png").resize((25,25)))
    ajouter_tache=tk.Button(centre_frame,text="Nouvelle tâche",bg="#222121",fg="white",image=icon1,compound="left",activebackground="#C5C1C1",highlightthickness=3,highlightbackground="#616164",command=ajouter_une_tache)
    ajouter_tache.pack(side=tk.TOP,anchor='e',padx=30,pady=30)
    ajouter_tache.image=icon1

    #Creer les menu (Bouttons): Liste de tâches ,Ajouter une tâche , Modifier une tâche , Supprimer une tâche
    icon1=ImageTk.PhotoImage(Image.open("list.png").resize((25,25)))
    liste_taches = tk.Button(menu_frame,text="Liste de tâches",fg="white",bg="#222121",image=icon1,compound="left",anchor="w",bd=0,relief="groove",highlightthickness=0,activebackground="#C5C1C1",height=40,command=lambda:liste_de_taches(centre_frame))
    liste_taches.pack(fill=tk.X,pady=0)
    icon2=ImageTk.PhotoImage(Image.open("add.png").resize((25,25)))
    ajouter_tache=tk.Button(menu_frame,text="Ajouter une tâche",fg="white",bg="#222121",image=icon2,compound="left",anchor="w",bd=0,relief="groove",highlightthickness=0,activebackground="#C5C1C1",height=40,command=ajouter_une_tache)
    ajouter_tache.pack(fill=tk.X,pady=0)
    icon3=ImageTk.PhotoImage(Image.open("edit.png").resize((25,25)))
    modifier_tache=tk.Button(menu_frame,text="Modifier une tâche",fg="white",bg="#222121",image=icon3,compound="left",anchor="w",bd=0,relief="groove",highlightthickness=0,activebackground="#C5C1C1",height=40,command=modifier_une_tache)
    modifier_tache.pack(fill=tk.X,pady=0)
    icon4=ImageTk.PhotoImage(Image.open("delete.png").resize((25,25)))
    supprimer_tache=tk.Button(menu_frame,text="Supprimer une tâche",fg="white",bg="#222121",image=icon4,compound="left",anchor="w",bd=0,relief="groove",highlightthickness=0,activebackground="#C5C1C1",height=40,command=supprimer_une_tache)
    supprimer_tache.pack(fill=tk.X,pady=0)



    liste_de_taches(centre_frame)

    fenetre.mainloop()

menu_principale()

File: task_scheduler.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

//Structure utiliser 
typedef struct tache
{
    char nom[100];
    char type_planification[50];
    char date[15];
    char heure[8];
    int jour_semaine;
    int jour_mois;
} tache;

//Declaration des fonctions
void menu_principal();
void  ajouter_une_tache();
void liste_taches(int utilisation);
tache supprimer_une_tache(int utilisation);
void modifier_une_tache();
int nombre_taches();

//Afficher le menu principal et demander à l'utilisateur de faire un choix
void menu_principal()
{
     int choix;
    printf("\033[10C \033[1;106;30m  MENU PRINCIPALE  \033[0m\n\n");
    printf("\033[10C\033[1;32m[1]\033[1;37m Ajouter une tâche \033[0m\n");
    printf("\033[10C\033[1;32m[2]\033[1;37m Tous les tâches \033[0m\n");
    printf("\033[10C\033[1;32m[3]\033[1;37m Modifier une tâche \033[0m\n");
    printf("\033[10C\033[1;32m[4]\033[1;37m Supprimer une tâche \033[0m\n");
    printf("\033[10C\033[1;32m[5]\033[1;37m Quitter \033[0m\n");
    printf("\033[10C\033[3;36mEntrez votre choix : \033[0m");
    scanf("%d", &choix);
    if(choix==1){ ajouter_une_tache();}
    else if(choix==2){liste_taches(0);}
    else if(choix==3){modifier_une_tache();}
    else if (choix==4){supprimer_une_tache(2);}
    else if (choix==5){exit(0);}
    else{menu_principal();}
}

//Cette demande au utilisateur la tâche avec la date d'exécution 
//Et puis enregistre cette dans un fichier "Liste_taches.csv" et dans le fichier cron (crontab -e)
void  ajouter_une_tache()
{
    int choix,i,min,heure,jour_du_mois,mois;
    char commande[255];
    char * type_planification [4] = {"Exécution unique","Tous les jours","Chaque semaine","Chaque mois"};
    tache nouvelle_tache ;
    //Alocation
    strcpy(nouvelle_tache.date,"00/00/0000");
    strcpy(nouvelle_tache.heure,"00:00");

    printf("\033[10C\033[1;106;30m  AJOUTER UNE TÂCHE  \033[0m\n\n");

    //Recuperer le type de planification de tâche
    printf("\033[10C\033[1;93mType de planification\033[0m\n");
    printf("\033[15C\033[1;32m[1]\033[1;37m Exécution unique \033[0m\n\033[15C\033[1;32m[2]\033[1;37m Tous les jours \033[0m\n\033[15C\033[1;32m[3]\033[1;37m Chaque semaine \033[0m\n\033[15C\033[1;32m[4]\033[1;37m Chaque mois \033[0m\n");
    printf("\033[15C\033[1;32m[5]\033[1;37m Retourner \033[0m\n\033[15C\033[1;32m[6]\033[1;37m Quitter \033[0m\n");
    printf("\033[10C\033[3;36mEntrez votre choix : \033[0m");
    scanf("%d", &choix);
    for (i=0;i<6;i++)
    {
        if(choix==i+1)
        {
            if(choix==5){ }
            else if(choix==6){exit(0);}
            else{strcpy(nouvelle_tache.type_planification,type_planification[i]);}
        }
    }
    //Recuperer la tâche à exécuter
    printf("\033[10C\033[1;93mTâche à exécuter : \033[0m");
    while(getchar() != '\n');  
    fgets(nouvelle_tache.nom,100,stdin);
    nouvelle_tache.nom[strcspn(nouvelle_tache.nom, "\n")] = '\0';
    //Recuperer la date et heure d'exécution d'une tâche
    if(strcmp(nouvelle_tache.type_planification,type_planification[0]) == 0)
    {
        printf("\033[10C\033[1;93mDate d'exécution [jj/mm/aaaa] : \033[0m");
        scanf("%s",nouvelle_tache.date);
    }
    if(strcmp(nouvelle_tache.type_planification,type_planification[3]) == 0)
    {
        printf("\033[10C\033[1;93mJour d'exécution chaque mois [1-31] : \033[0m");
        scanf("%d",&nouvelle_tache.jour_mois);
    }
    printf("\033[10C\033[1;93mHeure d'exécution [hh:mm] : \033[0m");
    scanf("%s",nouvelle_tache.heure);
    if(strcmp(nouvelle_tache.type_planification,type_planification[2]) == 0)
    {
        printf("\033[10C\033[1;93mJour de la semaine : \033[0m\n");
        printf("\033[15C\033[1;32m[0]\033[1;37m Dimanche \033[0m\n\033[15C\033[1;32m[1]\033[1;37m Lundi \033[0m\n\033[15C\033[1;32m[2]\033[1;37m Mardi \033[0m\n");
        printf("\033[15C\033[1;32m[3]\033[1;37m Mercredi \033[0m\n\033[15C\033[1;32m[4]\033[1;37m Jeudi \033[0m\n\033[15C\033[1;32m[5]\033[1;37m Vendredi \033[0m\n\033[15C\033[1;32m[6]\033[1;37m Samedi \033[0m\n");
        printf("\033[10C\033[3;36mEntrez votre choix : \033[0m");
        scanf("%d",&nouvelle_tache.jour_semaine);
    }

    //Ajouter la tâche dans le fichier "Liste_taches.csv" de la forme : Tâche,Type de planification,date,heure,joud de la semaine,jour du mois
    sprintf(commande,"echo \"%s,%s,%s,%s,%d,%d\" >> Liste_taches.csv",nouvelle_tache.nom,nouvelle_tache.type_planification,nouvelle_tache.date,nouvelle_tache.heure,nouvelle_tache.jour_semaine,nouvelle_tache.jour_mois);
    system(commande);

    //Recuperer : minute,heure,mois
    sscanf(nouvelle_tache.heure,"%d:%d",&heure,&min);
    sscanf(nouvelle_tache.date,"%d/%d/%*d",&jour_du_mois,&mois);

    //Ajouter une tâche cron 
    if(strcmp(nouvelle_tache.type_planification,type_planification[0]) == 0){sprintf(commande, " echo \"%d %d %d %d * %s\" >> temp_cron ", min, heure,jour_du_mois , mois, nouvelle_tache.nom);}
    else if(strcmp(nouvelle_tache.type_planification,type_planification[1]) == 0){sprintf(commande, " echo \"%d %d * * * %s\" >> temp_cron ", min, heure,nouvelle_tache.nom);}
    else if(strcmp(nouvelle_tache.type_planification,type_planification[2]) == 0){sprintf(commande, " echo \"%d %d * * %d %s\" >> temp_cron ", min, heure,nouvelle_tache.jour_semaine, nouvelle_tache.nom);}
    else if(strcmp(nouvelle_tache.type_planification,type_planification[3]) == 0){sprintf(commande, " echo \"%d %d %d * * %s\" >> temp_cron ", min, heure,jour_du_mois, nouvelle_tache.nom);}
    system(commande);
    sprintf(commande, "cat temp_cron | crontab -");
    system(commande);
}

//Cette fonction permet de compter le nombre de tâches dans le fichier Liste_taches.csv
int nombre_taches()
{
    int compter ;
    char * ligne = malloc(255*sizeof(char));
    FILE *fichier = fopen("Liste_taches.csv", "r");
    compter = 0;
    if (fichier != NULL)
    {
        while (fgets(ligne, 255, fichier) != NULL) 
        {
            compter++;
        }
        fclose(fichier);
    }
    return (compter-1);
}


void liste_taches(int utilisation)
{
    int nombreTaches,i;
    char * ligne = malloc(255*sizeof(char));
    char * jour_de_semaine[7]={"Dimanche","Lundi","Mardi","Mercredi","Jeudi","Vendredi","Samedi"};
    char * type_planification [4] = {"Exécution unique","Tous les jours","Chaque semaine","Chaque mois"};
    tache tache1;
    FILE *fichier = fopen("Liste_taches.csv", "r");
    nombreTaches = nombre_taches();
    if (fichier != NULL) 
    {
        //Afficher le titre selon l'utilisation : LISTES , MODIFICATION , SUPPRESSION
        if(utilisation==0){printf("\033[10C\033[1;106;30m  TOUTES LES TÂCHES  \033[0m\n\n");}
        else if(utilisation==1){printf("\033[10C\033[1;106;30m  MODIFIER UNE TÂCHE  \033[0m\n\n");}
        else if(utilisation==2){printf("\033[10C\033[1;106;30m  SUPPRIMER UNE TÂCHE  \033[0m\n\n");}
        
        printf("\033[10C\033[1;32m[Numero de tâche]\033[1;93m\tTâche\t  \033[0;3;46mType de planification\033[3;0m (Date d'exécution)\033[0m\n\n");

        //Afficher toutes les tâches dans le fichier "Liste_taches.csv": qui contient toutes les tâches cron
        for (i=0;i<nombreTaches;i++)
        {
            fgets(ligne,255,fichier);
            sscanf(ligne, "%[^,],%[^,],%[^,],%[^,],%d,%d", tache1.nom, tache1.type_planification, tache1.date, tache1.heure, &tache1.jour_semaine, &tache1.jour_mois);
            
            if(strcmp(tache1.type_planification,type_planification[0])==0){printf("\033[10C\033[1;32m[%d]\033[1;93m\t%s \n\033[10C\t  \033[0;3;46m%s\033[3;0m (%s à %s)\033[0m\n\n",i,tache1.nom, tache1.type_planification, tache1.date, tache1.heure);}
            else if(strcmp(tache1.type_planification,type_planification[1])==0){printf("\033[10C\033[1;32m[%d]\033[1;93m\t%s \n\033[10C\t  \033[0;3;44m%s\033[3;0m ( à %s)\033[0m\n\n",i,tache1.nom, tache1.type_planification, tache1.heure);}
            else if(strcmp(tache1.type_planification,type_planification[2])==0){printf("\033[10C\033[1;32m[%d]\033[1;93m\t%s \n\033[10C\t  \033[0;3;45m%s\033[3;0m (tous le %s à %s)\033[0m\n\n",i,tache1.nom, tache1.type_planification, jour_de_semaine[tache1.jour_semaine], tache1.heure);}
            else if(strcmp(tache1.type_planification,type_planification[3])==0){printf("\033[10C\033[1;32m[%d]\033[1;93m\t%s \n\033[10C\t  \033[0;3;42m%s\033[3;0m (Tous le %dème du mois à %s)\033[0m\n\n",i,tache1.nom, tache1.type_planification, tache1.jour_mois, tache1.heure);}
        }
        fclose(fichier);
    }
}

//Cette fonction permet de supprimer une tâche dans crontab -e
tache supprimer_une_tache(int utilisation)
{
    int choix,i;
    char ligne[255];
    char commade[255];
    tache tache_supprimer;
    FILE *fichier_src = fopen("Liste_taches.csv", "r");
    FILE *tmp = fopen("temp.csv", "w");
    liste_taches(utilisation);        
    printf("\033[10C\033[3;36mEntrez votre choix : \033[0m");
    scanf("%d",&choix);
    if(fichier_src!=NULL && tmp!=NULL)
    {
        i=0;
        while (fgets(ligne, sizeof(ligne), fichier_src))
        {
            if(i!=choix){fputs(ligne, tmp); } // Copier toutes les lignes sauf celle à supprimer
            else{sscanf(ligne, "%[^,],%[^,],%[^,],%[^,],%d,%d",tache_supprimer.nom,tache_supprimer.type_planification,tache_supprimer.date,tache_supprimer.heure,&tache_supprimer.jour_semaine,&tache_supprimer.jour_mois);}
            i++;
        }
        fclose(fichier_src);
        fclose(tmp);
    }
    // Remplacer le fichier original par le fichier temporaire
    remove("Liste_taches.csv");
    rename("temp.csv", "Liste_taches.csv");
    sprintf(commade,"crontab -l | grep -v \"%s\" | crontab -",tache_supprimer.nom);
    system(commade);
    return(tache_supprimer);
}

//Cette fonction permet de modifier une tâche 
void modifier_une_tache()
{
    tache tache_a_modifier;
    //Cette fonction demande au utilisateur le numero de tâche à modifier puis supprimer l'ancienne tâche 
    tache_a_modifier=supprimer_une_tache(1);
    //Afficher les ancienne tâche 
    printf("\033[10C\033[1;106;30m  ANCIENNE TÂCHE  \033[0m\n\n");
    printf("\033[10C\033[1;93mTâche à exécuter : \033[3;37m %s \033[0m",tache_a_modifier.nom);
    printf("\033[10C\033[1;93mType de planification :\033[3;37m %s \033[0m\n",tache_a_modifier.type_planification);
    printf("\033[10C\033[1;93mDate d'exécution [jj/mm/aaaa] :\033[3;37m %s \033[0m",tache_a_modifier.date);
    printf("\033[10C\033[1;93mHeure d'exécution [hh:mm] : \033[3;37m %s\033[0m",tache_a_modifier.heure);
    printf("\033[10C\033[1;93mJour d'exécution chaque mois [1-31] :\033[3;37m %d \033[0m",tache_a_modifier.jour_mois);
    printf("\033[10C\033[1;93mJour de la semaine : \033[3;37m %d\033[0m\n",tache_a_modifier.jour_semaine);
    ajouter_une_tache();
}

int main()
{
   menu_principal();
}